nova/quota.py

```diff
@@ -1093,19 +1093,47 @@ def update_resources_from_quota_classes(self):
             nova_context.get_admin_context(), 'flavors')
         del flavor_quotas['class_name']
 
-        resources = {}
         for key, value in flavor_quotas.items():
             resources[key] = CountableResource(key,
                 _instances_cores_ram_count, default=value)
 
-        resources.update(self._original_resources)
+    def update_resources_from_hardware_versions(self, resources):
+        """Update the resources dict to contain cores/ram for hw versions"""
 
-        self._resources = resources
+    def update_resources_from_flavors(self, resources):
+        """Update the resources dict to contain flavor based resources
+
+        Flavors can contain properties in their extra_specs which define that
+        we need custom quota resources for them e.g. separate instances quota
+        or separate hardware versioned cores/ram quota.
+        """
+        additional_resources = set()
+        ctx = nova_context.get_admin_context()
+        for flavor in objects.FlavorList.get_all(ctx):
+            extra_specs = flavor['extra_specs']
+
+            if extra_specs.get('quota:separate') == 'true':
+                additional_resources.add(f"instances_{flavor.name}")
+
+            if hw_version := extra_specs.get('quota:hw_version'):
+                additional_resources.add(f"cores_{hw_version}")
+                additional_resources.add(f"ram_{hw_version}")
+
+        for resource in additional_resources:
+            resources[resource] = CountableResource(resource,
+                _instances_cores_ram_count, default=0)
 
     @property
     def resources(self):
         if self._resources is None:
```

Comment:
Wouldn't it make sense to do here some refresh every X minutes, like you did with the project tags?

Reply:
Yes, I would definitely want to change to that, was was just not sure how to do it, yet.